app/cloud.go

```diff
@@ -18,6 +21,59 @@ func (a *App) getSysAdminsEmailRecipients() ([]*model.User, *model.AppError) {
 	return a.GetUsers(userOptions)
 }
 
+// SendAdminUpgradeRequestEmail takes the username of user trying to alert admins and then applies rate limit of n (number of admins) emails per user per day
+// before sending the emails.
+func (a *App) SendAdminUpgradeRequestEmail(username string, subscription *model.Subscription) *model.AppError {
+	if a.Srv().License() == nil || (a.Srv().License() != nil && !*a.Srv().License().Features.Cloud) {
+		return nil
+	}
+
+	if subscription != nil && subscription.IsPaidTier == "true" {
+		return nil
+	}
+
+	if a.Srv().EmailService.PerDayEmailRateLimiter == nil {
+		return model.NewAppError("app.SendAdminUpgradeRequestEmail", "app.email.no_rate_limiter.app_error", nil, fmt.Sprintf("username=%s", username), http.StatusInternalServerError)
+	}
+
+	// rate limit based on username as key
+	rateLimited, result, err := a.Srv().EmailService.PerDayEmailRateLimiter.RateLimit(username, 1)
+	if err != nil {
+		return model.NewAppError("app.SendAdminUpgradeRequestEmail", "app.email.setup_rate_limiter.app_error", nil, fmt.Sprintf("username=%s, error=%v", username, err), http.StatusInternalServerError)
+	}
+
+	if rateLimited {
+		return model.NewAppError("app.SendAdminUpgradeRequestEmail",
+			"app.email.rate_limit_exceeded.app_error", map[string]interface{}{"RetryAfter": result.RetryAfter.String(), "ResetAfter": result.ResetAfter.String()},
+			fmt.Sprintf("username=%s, retry_after_secs=%f, reset_after_secs=%f",
+				username, result.RetryAfter.Seconds(), result.ResetAfter.Seconds()),
+			http.StatusRequestEntityTooLarge)
```

Comment:
Probably here fits better 429 instead of 413

Reply:
I believe this is Agnivas design for email rate limit errors: https://github.com/mattermost/mattermost-server/pull/15230

Reply:
^ @agnivade about the reason of using 413 instead of 429 for rate limiting?

Reply:
It's explained in the PR body ;)

> The status code is chosen as 413 (entity too large) instead of 429 (too many requests)
because it's not the request which is rate limited, but the payload inside it which is.

Reply:
See https://github.com/mattermost/mattermost-server/pull/15230#issue-466523280 for a detailed context.

Reply:
Noted! We can keep it as it is but one thing worth mentioning is that we also rate limiting by the number of emails sent (requests) so I kind of a mixed scenario

Reply:
@AGMETEOR can you revert your changes as you originally had them returning `http.StatusRequestEntityTooLarge` per Agniva's design.

Reply:
> We can keep it as it is but one thing worth mentioning is that we also rate limiting by the number of emails sent (requests) so I kind of a mixed scenario

Yep, I explain it in more detail in the PR message. But essentially, it is rate limiting within rate limiting. There is no such response code to indicate that. I wanted to keep 429 purely to indicate rate limiting errors by the main rate limiter, and not by anything else internally. And returning a 429 error is also accompanied by `X-Retry-After` and `X-Reset-After`. We _could_ do `X-Email-Invite-Reset-After`, but I decided against it to keep the API clean.

Reply:
Understood, thank you for the explanation Agniva :)